File: src/main/java/org/xyp/project/todoapp/core/txop/TransactionalOp.java
package org.xyp.project.todoapp.core.txop;

import java.util.function.Supplier;

public interface TransactionalOp {
    <T> T returnInTx(Supplier<T> supplier);

    <T> T returnInNewTx(Supplier<T> supplier);

    void runInTx(Runnable runner);

    void runInNewTx(Runnable runner);
}

File: src/main/kotlin/org/xyp/todoapp/core/txop/TransactionalOp.kt
package org.xyp.todoapp.core.txop

interface TransactionalOp {
    fun <T> returnInTx(supplier: () -> T?): T?

    fun <T> returnInNewTx(supplier: () -> T?): T?

    fun runInTx(runner: () -> Unit)

    fun runInNewTx(runner: () -> Unit)
}